gmail/gmail_connector.bal

```diff
@@ -0,0 +1,723 @@
+    encodedRequest = encodedRequest.replace("+", "-").replace("/", "_");
+    //Set the encoded message as raw
+    message.raw = encodedRequest;
+    http:Request request = {};
+    http:Response response = {};
+    http:HttpConnectorError connectionError = {};
+    GmailError gmailError = {};
+    json jsonPayload = {"raw":message.raw};
+    string sendMessagePath = USER_RESOURCE + userId + MESSAGE_SEND_RESOURCE;
+    request.setJsonPayload(jsonPayload);
+    request.setHeader(CONTENT_TYPE, APPLICATION_JSON);
+    var sendMessageResponse = oauthEP -> post(sendMessagePath, request);
+    match sendMessageResponse {
+        http:Response res => response = res;
+        http:HttpConnectorError connectErr => connectionError = connectErr;
+    }
+    if (connectionError.message != "") {
+        gmailError.errorMessage = connectionError.message;
+        gmailError.statusCode = connectionError.statusCode;
+        return gmailError;
+    }
+    json jsonSendMessageResponse;
+    match response.getJsonPayload() {
+        mime:EntityError err => gmailError.errorMessage = err.message;
+        json jsonResponse => jsonSendMessageResponse = jsonResponse;
+    }
+    if (gmailError.errorMessage != "") {
+        return gmailError;
+    }
+    if (response.statusCode == STATUS_CODE_200_OK) {
+        string id = jsonSendMessageResponse.id.toString();
+        string threadId = jsonSendMessageResponse.threadId.toString();
+        return (id, threadId);
+    } else {
+        gmailError.errorMessage = jsonSendMessageResponse.error.message.toString();
+        gmailError.statusCode = response.statusCode;
+        return gmailError;
+    }
+}
+
+@Description {value:"Read the specified mail from users mailbox"}
+@Param {value:"userId: user's email address. The special value -> me"}
+@Param {value:"messageId: message id of the specified mail to retrieve"}
+@Param {value:"filter: GetMessageThreadFilter struct object with the optional format and metadataHeaders query parameters"}
+@Param {value:"Returns the specified mail as a Message struct"}
+@Return {value:"Returns GmailError if the message cannot be read successfully"}
+public function <GmailConnector gmailConnector> readMail (string userId, string messageId, GetMessageThreadFilter filter) returns (Message)|GmailError {
+    endpoint oauth2:OAuth2Endpoint oauthEP = gmailConnector.oauthEndpoint;
+    http:Request request = {};
+    http:Response response = {};
+    http:HttpConnectorError connectionError = {};
+    GmailError gmailError = {};
+    string uriParams = "";
+    string readMailPath = USER_RESOURCE + userId + MESSAGE_RESOURCE + "/" + messageId;
+    //Add format optional query parameter
+    uriParams = filter.format != "" ? uriParams + FORMAT + filter.format : uriParams;
+    //Add the optional meta data headers as query parameters
+    foreach metaDataHeader in filter.metadataHeaders {
+        uriParams = metaDataHeader != "" ? uriParams + METADATA_HEADERS + metaDataHeader : uriParams;
```

Comment:
Don't you need to add "&" character when concatenating more uri parameters with the existing uri parameters?